compiler/src/main/scala/org/scalajs/core/compiler/JSEncoding.scala

```diff
@@ -99,12 +105,26 @@ trait JSEncoding extends SubComponent { self: GenJSCode =>
      * because they are emitted as private by our .scala source files, but
      * they are considered public at use site since their symbols come from
      * Java-emitted .class files.
+     *
+     * Starting with 2.12.0-RC2, we also special case outer fields. This
+     * essentially fix #2382, which is caused by a class having various $outer
+     * pointers in its hierarchy that points to different outer instances.
+     * Without this fix, they all collapse to the same field in the IR. We
+     * cannot fix this for all Scala versions at the moment, because that would
+     * break backwards binary compatibility. We *do* fix it for 2.12.0-RC2
+     * onwards because that also fixes #2625, which surfaced in 2.12 and is
+     * therefore a regression. We can do this because the 2.12 ecosystem is
+     * not binary compatible anyway (because of Scala) so we can break it on
+     * our side at the same time.
      */
-    val idSuffix =
-      if (sym.isPrivate || allRefClasses.contains(sym.owner))
+    val idSuffix: String = {
+      if (applyFixFor2382 && sym.isOuterField)
+        encodeClassFullName(sym.owner)
+      else if (sym.isPrivate || allRefClasses.contains(sym.owner))
         sym.owner.ancestors.count(!_.isTraitOrInterface).toString
```

Comment:
Well ... the problem is that scalac generates *protected* fields for `$outer`. And that's basically the only way to get scalac to ever generate a protected field. We see that after mixin:
```scala
scala> class Foo { class Bar }
[[syntax trees at end of                     mixin]] // <console>
package $line3 {
  ...
  class iw$Foo$Bar extends Object {
    <synthetic> <paramaccessor> <artifact> protected val $outer: $line3.iw$Foo = _;
    <synthetic> <stable> <artifact> def $outer(): $line3.iw$Foo = iw$Foo$Bar.this.$outer;
    def <init>($outer: $line3.iw$Foo): $line3.iw$Foo$Bar = {
      if ($outer.eq(null))
        throw null
      else
        iw$Foo$Bar.this.$outer = $outer;
      iw$Foo$Bar.super.<init>();
      ()
    }
  }
}
```
And apparently, the JVM is fine with this. It seems to disambiguate fields by their enclosing class. Field selection opcodes reference an enclosing class in addition to a field name (and actually the field *type* as well, I've been told; so you can *overload* a field!!), which allows it to disambiguate outer fields correctly. According to a small test with `javac`, I figured that it also considers "inherited" fields (so if it says `B.x` but `x` is defined in `A` that `B` extends, it will find it).

Alternatively to `sym.isOuterField`, we could probably do `sym.isProtected`. I wager it would identify the exact same set of symbols.

We could be much more extreme, and do this for *all* symbols, including public fields. That would be valid if moving a field one class up the hierarchy is considered a binary *incompatible* change. But that's not true on the JVM according to my experiments. It would be OK for all Scala-generated code, though, since only synthetic fields are ever non-private. So maybe we *could* declare it binary incompatible for .sjsir.